src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/AppleCCCryptor.cs

```diff
@@ -39,46 +39,36 @@ protected override void Dispose(bool disposing)
         {
-            Debug.Assert(input != null, "Expected valid input, got null");
-            Debug.Assert(inputOffset >= 0, $"Expected non-negative inputOffset, got {inputOffset}");
-            Debug.Assert(count > 0, $"Expected positive count, got {count}");
-            Debug.Assert((count % BlockSizeInBytes) == 0, $"Expected count aligned to block size {BlockSizeInBytes}, got {count}");
-            Debug.Assert(input.Length - inputOffset >= count, $"Expected valid input length/offset/count triplet, got {input.Length}/{inputOffset}/{count}");
-            Debug.Assert(output != null, "Expected valid output, got null");
-            Debug.Assert(outputOffset >= 0, $"Expected non-negative outputOffset, got {outputOffset}");
-            Debug.Assert(output.Length - outputOffset >= count, $"Expected valid output length/offset/count triplet, got {output.Length}/{outputOffset}/{count}");
-
-            return CipherUpdate(input, inputOffset, count, output, outputOffset);
+            Debug.Assert(input.Length > 0);
+            Debug.Assert((input.Length % BlockSizeInBytes) == 0);
+
+            return CipherUpdate(input, output);
         }
 
-        public override byte[] TransformFinal(byte[] input, int inputOffset, int count)
+        public override int TransformFinal(ReadOnlySpan<byte> input, Span<byte> output)
         {
-            Debug.Assert(input != null, "Expected valid input, got null");
-            Debug.Assert(inputOffset >= 0, $"Expected non-negative inputOffset, got {inputOffset}");
-            Debug.Assert(count >= 0, $"Expected non-negative count, got {count}");
-            Debug.Assert((count % BlockSizeInBytes) == 0, $"Expected count aligned to block size {BlockSizeInBytes}, got {count}");
-            Debug.Assert(input.Length - inputOffset >= count, $"Expected valid input length/offset/count triplet, got {input.Length}/{inputOffset}/{count}");
+            Debug.Assert((input.Length % BlockSizeInBytes) == 0);
+            Debug.Assert(input.Length <= output.Length);
 
-            byte[] output = ProcessFinalBlock(input, inputOffset, count);
+            int written = ProcessFinalBlock(input, output);
             Reset();
-            return output;
+            return written;
         }
 
-        private unsafe byte[] ProcessFinalBlock(byte[] input, int inputOffset, int count)
+        private unsafe int ProcessFinalBlock(ReadOnlySpan<byte> input, Span<byte> output)
         {
-            if (count == 0)
+            if (input.Length == 0)
             {
-                return Array.Empty<byte>();
+                return 0;
             }
 
-            byte[] output = new byte[count];
-            int outputBytes = CipherUpdate(input, inputOffset, count, output, 0);
+            int outputBytes = CipherUpdate(input, output);
             int ret;
             int errorCode;
 
-            fixed (byte* outputStart = &output[0])
+            fixed (byte* outputStart = output)
```

Comment:
Is it possible for us to get here with `output` being empty? If so, `outputStart` is 0/NULL, which might violate requirements of the called methods.

Reply:
It's not expected. This is asserted, just not straight forwardly.

A few lines up, we return early if `input.Length` is 0.

```
if (input.Length == 0)
{
    return 0;
}
```

So we know that input length is > 0. `ProcessFinalBlock` is called in exactly one place, by `TransformFinal`. It asserts:

```
Debug.Assert(input.Length <= output.Length);
```

So output's length by the time it reaches the `fixed` is `1 <= output.Length`.

I think I'll add an assert here anyway to spell all that out.